File: C#/Silverlight/source/LSilverlight-0.3.3/Loon.Utils.Xml/XMLElement.cs
namespace Loon.Utils.Xml {

    using System.Collections.Generic;
	using Loon.Core;
    using Loon.Java.Collections;
   
	public class XMLElement : LRelease {
	
		private string name;
	
		private IDictionary<string, XMLAttribute> attributes;
	
		private IList<object> contents;
	
		private XMLElement parent;
	
		internal XMLElement(string n) {
            this.attributes = new Dictionary<string, XMLAttribute>();
            this.contents = new List<object>();
			this.name = n;
		}
	
		public XMLAttribute GetAttribute(string n) {
			if (!this.attributes.ContainsKey(n))
				throw new System.Exception("Unknown attribute name '" + n
						+ "' in element '" + this.name + "' !");
            return attributes[n];
		}
	
		public string GetAttribute(string n, string v) {
            if (!this.attributes.ContainsKey(n))
            {
				return v;
			}
            return attributes[n].GetValue();
		}
	
		public int GetIntAttribute(string n, int v) {
            if (!this.attributes.ContainsKey(n))
            {
				return v;
			}
            return attributes[n].GetIntValue();
		}
	
		public float GetFloatAttribute(string n, float v) {
			if (!this.attributes.ContainsKey(n)) {
				return v;
			}
            return attributes[n].GetFloatValue();
		}
	
		public double GetDoubleAttribute(string n, double v) {
			if (!this.attributes.ContainsKey(n)) {
				return v;
			}
            return attributes[n].GetDoubleValue(); ;
		}
	
		public bool GetBoolAttribute(string n, bool v) {
			if (!this.attributes.ContainsKey(n)) {
				return v;
			}
            return attributes[n].GetBoolValue();
		}

        public IDictionary<string, XMLAttribute> GetAttributes()
        {
			return this.attributes;
		}
	
		public bool HasAttribute(string n) {
			return this.attributes.ContainsKey(n);
		}
	
		public IIterator Elements() {
			return new IteratorAdapter(this.contents.GetEnumerator());
		}
	
		public XMLElement GetFirstChild() {
			for (IIterator e = Elements(); e.HasNext();) {
				object o = e.Next();
				if (!(o  is  XMLElement)) {
					continue;
				}
				return (XMLElement) o;
			}
			return null;
		}
	
		public XMLElement GetChildrenByName(string n) {
			for (IIterator e = Elements(); e.HasNext();) {
				object o = e.Next();
				if ((!(o  is  XMLElement))
						|| (!((XMLElement) o).GetName().Equals(n))) {
					continue;
				}
				return (XMLElement) o;
			}
			return null;
		}
	
		public List<XMLElement> Find(string n) {
			List<XMLElement> v = new List<XMLElement>();
			for (IIterator e = Elements(); e.HasNext();) {
				object o = e.Next();
				if ((!(o  is  XMLElement))) {
					continue;
				}
				XMLElement ele = (XMLElement) o;
				if (!ele.Equals(ele.GetName())) {
					IIterator it = ele.Elements(n);
					for (; it.HasNext();) {
						XMLElement child = (XMLElement) it.Next();
						child.parent = ele;
                        CollectionUtils.Add(v, child);
					}
					continue;
				} else if (ele.Equals(ele.GetName())) {
                    CollectionUtils.Add(v, (XMLElement)o);
					continue;
				}
			}
			return v;
		}
	
		public List<XMLElement> List(string n) {
			List<XMLElement> v = new List<XMLElement>();
			for (IIterator e = Elements(); e.HasNext();) {
				object o = e.Next();
				if ((!(o  is  XMLElement))
						|| (!((XMLElement) o).GetName().Equals(n))) {
					continue;
				}
                CollectionUtils.Add(v, (XMLElement)o);
			}
			return v;
		}

        public IIterator Elements(string n)
        {
			List<object> v = new List<object>();
			for (IIterator e = Elements(); e.HasNext();) {
				object o = e.Next();
				if ((!(o  is  XMLElement))
						|| (!((XMLElement) o).GetName().Equals(n))) {
					continue;
				}
                CollectionUtils.Add(v, o);
			}
			return new IteratorAdapter(v.GetEnumerator());
		}
	
		public void AddAllTo(List<XMLElement> list) {
			for (IIterator e = Elements(); e.HasNext();) {
				object o = e.Next();
				if ((!(o  is  XMLElement))
						|| (!((XMLElement) o).GetName().Equals(name))) {
					continue;
				}
                CollectionUtils.Add(list, (XMLElement)o);
			}
		}
	
		public string GetName() {
			return this.name;
		}
	
		public XMLElement GetParent() {
			return this.parent;
		}
	
		public string GetContents() {
			System.Text.StringBuilder sbr = new System.Text.StringBuilder(1024);
			for (IIterator e = Elements(); e.HasNext();) {
				sbr.Append(e.Next().ToString());
			}
			return sbr.ToString();
		}
	
		public override string ToString() {
			ICollection<string> set = this.attributes.Keys;
			string str1 = "<" + this.name;
			for (IIterator it = new IteratorAdapter(set.GetEnumerator()); it.HasNext();) {
				string str2 = (string) it.Next();
				str1 = str1 + " " + str2 + " = \"" + GetAttribute(str2).GetValue()
						+ "\"";
			}
			str1 = str1 + ">";
			str1 = str1 + GetContents();
			str1 = str1 + "</" + this.name + ">";
			return str1;
		}
	
		internal XMLAttribute AddAttribute(string n, string value_ren) {
			XMLAttribute attribute = new XMLAttribute(n, value_ren);
            attributes.Add(n,attribute);
			return attribute;
		}
	
		internal void AddContents(object o) {
            contents.Add(o);
		}
	
		public virtual void Dispose() {
			if (attributes != null) {
				attributes.Clear();
				attributes = null;
			}
			if (contents != null) {
                contents.Clear();
				contents = null;
			}
		}
	
	}
}

File: Java/Loon-Neo/src/loon/utils/xml/XMLElement.java
/**
 * Copyright 2008 - 2011
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.1
 */
package loon.utils.xml;

import java.util.Iterator;

import loon.LSysException;
import loon.LSystem;
import loon.utils.Base64Coder;
import loon.utils.ObjectMap;
import loon.utils.TArray;

public class XMLElement {

	private String name;

	private ObjectMap<String, XMLAttribute> attributes;

	private TArray<Object> contents;

	private XMLElement parent;

	public XMLElement() {
		this(null);
	}

	public XMLElement(String name) {
		this.attributes = new ObjectMap<String, XMLAttribute>();
		this.contents = new TArray<Object>();
		this.name = name;
	}

	public byte[] readContentBinHex() {
		byte[] buffer = new byte[0x1000];
		readBinHex(buffer, 0, 0x1000);
		return buffer;
	}

	public int readBinHex(byte[] buffer, int offset, int length) {
		if (offset < 0) {
			throw new LSysException("Offset must be non-negative integer.");
		} else if (length < 0) {
			throw new LSysException("Length must be non-negative integer.");
		} else if (buffer.length < offset + length) {
			throw new LSysException("buffer length is smaller than the sum of offset and length.");
		}
		if (length == 0) {
			return 0;
		}
		char[] chars = new char[length * 2];
		int charsLength = readValueChunk(chars, 0, length * 2);
		return Base64Coder.fromBinHexString(chars, offset, charsLength, buffer);
	}

	private int readValueChunk(char[] buffer, int offset, int length) {
		StringBuffer textCache = new StringBuffer(length);
		for (Iterator<?> e = elements(); e.hasNext();) {
			textCache.append(e.next().toString());
		}
		int min = textCache.length();
		if (min > length) {
			min = length;
		}
		String str = textCache.substring(0, min);
		System.arraycopy(str.toCharArray(), offset, buffer, 0, length);
		if (min < length) {
			return min + readValueChunk(buffer, offset + min, length - min);
		} else {
			return min;
		}
	}

	public XMLAttribute getAttribute(String name) {
		if (!this.attributes.containsKey(name))
			throw new LSysException("Unknown attribute name '" + name + "' in element '" + this.name + "' !");
		return this.attributes.get(name);
	}

	public String getAttribute(String name, String v) {
		if (!this.attributes.containsKey(name)) {
			return v;
		}
		return (this.attributes.get(name)).getValue();
	}

	public int getIntAttribute(String name, int v) {
		if (!this.attributes.containsKey(name)) {
			return v;
		}
		return (this.attributes.get(name)).getIntValue();
	}

	public float getFloatAttribute(String name, float v) {
		if (!this.attributes.containsKey(name)) {
			return v;
		}
		return (this.attributes.get(name)).getFloatValue();
	}

	public double getDoubleAttribute(String name, double v) {
		if (!this.attributes.containsKey(name)) {
			return v;
		}
		return (this.attributes.get(name)).getDoubleValue();
	}

	public boolean getBoolAttribute(String name, boolean v) {
		if (!this.attributes.containsKey(name)) {
			return v;
		}
		return (this.attributes.get(name)).getBoolValue();
	}

	public ObjectMap<String, XMLAttribute> getAttributes() {
		return this.attributes;
	}

	public boolean hasAttribute(String name) {
		return this.attributes.containsKey(name);
	}

	public Iterator<Object> elements() {
		return this.contents.iterator();
	}
	
	public TArray<XMLElement> list() {
		TArray<XMLElement> lists = new TArray<XMLElement>(contents.size);
		for (Iterator<?> e = elements(); e.hasNext();) {
			Object o = e.next();
			if (!(o instanceof XMLElement)) {
				continue;
			}
			lists.add((XMLElement) o);
		}
		return lists;
	}

	public XMLElement getFirstChild() {
		for (Iterator<?> e = elements(); e.hasNext();) {
			Object o = e.next();
			if (!(o instanceof XMLElement)) {
				continue;
			}
			return (XMLElement) o;
		}
		return null;
	}

	public XMLElement getChildrenByName(String name) {
		for (Iterator<?> e = elements(); e.hasNext();) {
			Object o = e.next();
			if ((!(o instanceof XMLElement)) || (!((XMLElement) o).getName().equals(name))) {
				continue;
			}
			return (XMLElement) o;
		}
		return null;
	}

	public TArray<XMLElement> find(String name) {
		TArray<XMLElement> v = new TArray<XMLElement>();
		for (Iterator<?> e = elements(); e.hasNext();) {
			Object o = e.next();
			if ((!(o instanceof XMLElement))) {
				continue;
			}
			XMLElement ele = (XMLElement) o;
			if (!ele.equals(ele.getName())) {
				Iterator<?> it = ele.elements(name);
				for (; it.hasNext();) {
					XMLElement child = (XMLElement) it.next();
					child.parent = ele;
					v.add(child);
				}
				continue;
			} else if (ele.equals(ele.getName())) {
				v.add((XMLElement) o);
				continue;
			}
		}
		return v;
	}

	public TArray<XMLElement> list(String name) {
		TArray<XMLElement> v = new TArray<XMLElement>();
		for (Iterator<?> e = elements(); e.hasNext();) {
			Object o = e.next();
			if ((!(o instanceof XMLElement)) || (!((XMLElement) o).getName().equals(name))) {
				continue;
			}
			v.add((XMLElement) o);
		}
		return v;
	}

	public Iterator<?> elements(String name) {
		TArray<Object> v = new TArray<Object>();
		for (Iterator<?> e = elements(); e.hasNext();) {
			Object o = e.next();
			if ((!(o instanceof XMLElement)) || (!((XMLElement) o).getName().equals(name))) {
				continue;
			}
			v.add(o);
		}
		return v.iterator();
	}

	public void addAllTo(TArray<XMLElement> list) {
		for (Iterator<?> e = elements(); e.hasNext();) {
			Object o = e.next();
			if ((!(o instanceof XMLElement)) || (!((XMLElement) o).getName().equals(name))) {
				continue;
			}
			list.add((XMLElement) o);
		}
	}

	public String getName() {
		return this.name;
	}

	public XMLElement getParent() {
		return this.parent;
	}

	public String getContents() {
		StringBuffer sbr = new StringBuffer(1024);
		for (Iterator<?> e = elements(); e.hasNext();) {
			sbr.append(e.next().toString());
		}
		return sbr.toString();
	}

	public XMLAttribute addAttribute(String name, int value) {
		XMLAttribute attribute = new XMLAttribute(name, String.valueOf(value));
		this.attributes.put(name, attribute);
		return attribute;
	}

	public XMLAttribute addAttribute(String name, String value) {
		XMLAttribute attribute = new XMLAttribute(name, value);
		this.attributes.put(name, attribute);
		return attribute;
	}

	public XMLElement addContents(Object o) {
		this.contents.add(o);
		return this;
	}

	@Override
	public String toString() {
		StringBuffer builder = new StringBuffer();
		if (this.name == null) {
			builder.append(LSystem.EMPTY);
		} else {
			builder.append('<');
			builder.append(this.name);
		}
		for (String str2 : attributes.keys()) {
			builder.append(' ');
			builder.append(str2);
			builder.append(" = \"");
			builder.append(getAttribute(str2).getValue());
			builder.append("\"");
		}
		if (this.name != null) {
			builder.append(">\n");
		}
		builder.append(getContents());
		if (this.name != null) {
			builder.append("</" + this.name + ">\n");
		}
		return builder.toString();
	}

	public XMLElement dispose() {
		if (attributes != null) {
			attributes.clear();
			attributes = null;
		}
		if (contents != null) {
			contents.clear();
			contents = null;
		}
		return this;
	}

}
